subsys/net/lib/sockets/sockets_can.c

```diff
@@ -167,9 +168,7 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
 		}
 	}
 
-	if (clone && clone != pkt) {
```

Comment:
Why did you remove the clone variable, it is used if we have multiple socket_can receives?

Reply:
I will test this once more, but I was confident that `pkt` needed to be unref'ed otherwise the `net_pkt_` buffer will fill up and won't be cleared.

Reply:
The clone was important when we have multiple socket_can listeners. This can be tested with samples/net/sockets/can sample app and setting CONFIG_NET_SOCKETS_CAN_RECEIVERS=2

Reply:
I've tested it again. And I think I know what's wrong.
In my opinion the initial code is invalid.

I'm seeing the problem when only one receiver is configured.
And if it's configured with a filter which does **not**  matches the current packet.
The packet is not unreferenced.

If no receiver is found then `clone == NULL`  and if statement is ignored.
```
if (clone && clone != pkt) {
	net_pkt_unref(pkt);
}
```
So I think the statement should be:
```
if (!clone || clone != pkt) {
	net_pkt_unref(pkt);
}
```
If `clone` isn't set, this means no receiver is found. Then `pkt` has no value for the rest of the application and should not hold a place in the buffer because no one is waiting for this `pkt`
But if it has value, and it's cloned with `net_pkt_clone`. The orignal may be deleted, because it's cloned.

What do you think @jukkar?

Reply:
> I'm seeing the problem when only one receiver is configured.
> And if it's configured with a filter which does **not** matches the current packet.
> The packet is not unreferenced.
> 
> If no receiver is found then `clone == NULL` and if statement is ignored.
> 

There should be at least one receiver in this case as otherwise there would be no socket and this piece of code would not be called in the first place.

The original code tried to check whether there were other listeners for the same socket, and if there were, it cloned the pkt and delivered it to recipient. If there was more that one recipient, we cloned the packet and left the original pkt to be freed at the end of the func. This behaviour could be optimized so we would not need to unref the packet in this func.

Reply:
I think you don't understand. Let me start from the beginning.

It's not possible to filter on socketcan error frames in the can hardware driver.
Because it's a socketcan frame and not a can frame, its generated by the (inline) sockcan_generic module. Thats why it must be filtered in the net layer. 

If there are sockets binded to the interface, it's able to receive all frames or (almost) none.
An error frame does always trigger `zcan_received_cb`, if no receiver want's to receive this frame it must be unreferenced. We can't deny existence.

The original code relies on the hardware filter of the can driver. Only packages registered for are distributed to the contexts/sockets. Since the use of error frames, which are eventually mock frames to notify userspace, the `zcan_received_cb` is called from the socketcan_generic module (through `net_recv_data`) even if no package is received from the hardware driver. This frame is matched in `zcan_received_cb` (essentially this functionality was already present), if no one wants this frame it need to be unreferenced.

There were no excess frames in the original driver, because all was filtered by hardware. So all could be dispatched, and there was no direct problem. But if an excess frame is inserted through the socketcan_generic module. You would see the following error eventually:

 ` <err> can_driver: Failed to obtain RX buffer` from the socketcan_generic module.